src/main/java/com/arpnetworking/metrics/mad/PeriodWorker.java

```diff
@@ -16,70 +16,83 @@
-import akka.actor.Cancellable;
+import akka.actor.AbstractActorWithTimers;
+import akka.actor.ActorRef;
 import com.arpnetworking.logback.annotations.LogValue;
 import com.arpnetworking.metrics.incubator.PeriodicMetrics;
 import com.arpnetworking.metrics.mad.model.Record;
 import com.arpnetworking.steno.LogValueMapFactory;
 import com.arpnetworking.steno.Logger;
 import com.arpnetworking.steno.LoggerFactory;
+import com.arpnetworking.tsdcore.model.Key;
 import com.google.common.collect.Lists;
 
 import java.time.Duration;
 import java.time.Instant;
+import java.time.ZoneId;
 import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
 import java.util.List;
 import java.util.Map;
 import java.util.NavigableMap;
 import java.util.Optional;
 import java.util.TreeMap;
-import javax.annotation.Nullable;
 
 /**
  * Actor that aggregates a particular slice of the data set over time and metric.
  *
  * @author Ville Koskela (ville dot koskela at inscopemetrics dot io)
  */
-/* package private */ final class PeriodWorker extends AbstractActor {
+/* package private */ final class PeriodWorker extends AbstractActorWithTimers {
 
     /**
      * Public constructor. Since this is an {@code Actor} this method should not be
      * called directly, but instead you should use {@code Props}.
      */
     PeriodWorker(
+            final ActorRef aggregator,
+            final Key key,
             final Duration period,
+            final Duration idleTimeout,
             final Bucket.Builder bucketBuilder,
             final PeriodicMetrics periodicMetrics) {
+        _aggregator = aggregator;
+        _key = key;
         _period = period;
+        _idleTimeout = idleTimeout;
         _bucketBuilder = bucketBuilder;
         _periodicMetrics = periodicMetrics;
-        _hasRotateScheduled = false;
+        _hasReceivedRecords = false;
+        _nextScheduledRotationTime = Optional.empty();
     }
 
     @Override
     public void preStart() {
-        // TODO(ville): Schedule the actor to self-destruct if it doesn't receive any Record
-        // instances for some period of time (e.g. > the largest period). The caveat is
-        // that we need to deregister the actor from the Aggregator as well without
-        // creating a race condition. Note that period worker clean-up did not exist
-        // in the previous implementation either (not that it shouldn't; just one
-        // problem at a time).
         _periodicMetrics.recordCounter("actors/period_worker/started", 1);
+
+        timers().startPeriodicTimer(
+                IDLE_CHECK_TIMER,
+                IDLE_CHECK_MESSAGE,
+                _idleTimeout);
     }
 
     @Override
     public void postStop() {
-        if (_nextScheduledRotation != null) {
-            final boolean cancelResult = _nextScheduledRotation.cancel();
-            _nextScheduledRotation = null;
+        _periodicMetrics.recordCounter("actors/period_worker/stopped", 1);
+
+        timers().cancel(IDLE_CHECK_TIMER);
```

Comment:
You don't actually need to do this. But I think you should call super.postStop().

Reply:
Added super calls.